pkg/operator/controller/certificate/default_cert.go

```diff
@@ -17,67 +17,77 @@ import (
 
 // ensureDefaultCertificateForIngress creates or deletes an operator-generated
 // default certificate for a given IngressController as appropriate.  Returns true
-// if it creates, updates, or deletes the secret for the certificate, false
-// otherwise.
+// if it the secret exists, or false if it does not, as well as any errors.
 func (r *reconciler) ensureDefaultCertificateForIngress(caSecret *corev1.Secret, namespace string, deploymentRef metav1.OwnerReference, ci *operatorv1.IngressController) (bool, error) {
 	ca, err := crypto.GetCAFromBytes(caSecret.Data["tls.crt"], caSecret.Data["tls.key"])
 	if err != nil {
 		return false, fmt.Errorf("failed to get CA from secret %s/%s: %v", caSecret.Namespace, caSecret.Name, err)
 	}
-	desired, err := desiredRouterDefaultCertificateSecret(ca, namespace, deploymentRef, ci)
+	wantCert, desired, err := desiredRouterDefaultCertificateSecret(ca, namespace, deploymentRef, ci)
 	if err != nil {
 		return false, err
 	}
-	current, err := r.currentRouterDefaultCertificate(ci, namespace)
+	if !wantCert {
+		// If the operator generated certificate is not being used, ensure that the ingress controller's
+		// Spec.DefaultCertificate secret exists before deleting the operator generated secret.
+		// See https://bugzilla.redhat.com/show_bug.cgi?id=1887441
+		err := r.lookupUserSpecifiedRouterDefaultCertificate(ci, namespace)
+		if err != nil {
+			return false, fmt.Errorf("failed to lookup user specified default certificate: %v", err)
+		}
+	}
+
+	haveCert, current, err := r.currentRouterDefaultCertificate(ci, namespace)
 	if err != nil {
 		return false, err
 	}
 	switch {
-	case desired == nil && current == nil:
+	case !wantCert && !haveCert:
```

Comment:
I like the clearer variable names here, but maybe you can just add two lines to set wantCert if desired != nil and haveCert if current != nil, and keep the signatures/returns the same on the two methods you changed?

Reply:
Granted we've been slow to adopt the `haveFoo` boolean pattern across every `ensureFoo` function in our components, but we have been moving towards using it to essentially avoid lots of `desired != nil` checks. The functions in which I've made signature changes to belong to this file and are private, so I don't think changing them is a big deal. Personally, having the boolean return value is simpler for me to understand, especially in the more complicate `ensureFoo` functions, such as `ensureLoadBalancerService` (https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/load_balancer_service.go#L142)